lib/hanami/entity.rb

```diff
@@ -61,238 +51,27 @@ module Hanami
-      # @see Hanami::Model::Mapper
-      #
-      # @example
-      #   require 'hanami/model'
-      #
-      #   class User
-      #     include Hanami::Entity
-      #     attributes :name, :age
-      #   end
-      #   User.attributes => #<Set: {:id, :name, :age}>
-      #
-      # @example Given params is array of attributes
-      #   require 'hanami/model'
-      #
-      #   class User
-      #     include Hanami::Entity
-      #     attributes [:name, :age]
-      #   end
-      #   User.attributes => #<Set: {:id, :name, :age}>
-      #
-      # @example Extend entity
-      #   require 'hanami/model'
-      #
-      #   class User
-      #     include Hanami::Entity
-      #     attributes :name
-      #   end
-      #
-      #   class DeletedUser < User
-      #     include Hanami::Entity
-      #     attributes :deleted_at
-      #   end
-      #
-      #   User.attributes => #<Set: {:id, :name}>
-      #   DeletedUser.attributes => #<Set: {:id, :name, :deleted_at}>
-      #
-      def attributes(*attrs)
-        return @attributes ||= Set.new unless attrs.any?
-
-        Hanami::Utils::Kernel.Array(attrs).each do |attr|
-          if allowed_attribute_name?(attr)
-            define_attr_accessor(attr)
-            self.attributes << attr
-          end
-        end
-      end
-
-      # Define setter/getter methods for attributes.
-      #
-      # @param attr [Symbol] an attribute name
-      #
-      # @since 0.3.1
-      # @api private
-      def define_attr_accessor(attr)
-        attr_accessor(attr)
-      end
-
-      # Check if attr_reader define the given attribute
-      #
-      # @since 0.5.1
-      # @api private
-      def allowed_attribute_name?(name)
-        !instance_methods.include?(name)
-      end
-
-      protected
-
-      # @see Class#inherited
-      def inherited(subclass)
-        subclass.attributes(*attributes)
-        super
-      end
+    def initialize(attributes = nil)
+      @attributes = Utils::Hash.new((attributes || {}).dup).symbolize!
+      freeze
     end
 
-    # Defines a generic, inefficient initializer, in case that the attributes
-    # weren't explicitly defined with `.attributes=`.
-    #
-    # @param attributes [Hash] a set of attribute names and values
-    #
-    # @raise NoMethodError in case the given attributes are trying to set unknown
-    #   or private methods.
-    #
-    # @since 0.1.0
-    #
-    # @see .attributes
-    def initialize(attributes = {})
-      attributes.each do |k, v|
-        setter = "#{ k }="
-        public_send(setter, v) if respond_to?(setter)
-      end
+    def method_missing(m)
+      attributes.fetch(m, nil)
     end
 
-    # Overrides the equality Ruby operator
-    #
-    # Two entities are considered equal if they are instances of the same class
-    # and if they have the same #id.
-    #
-    # @since 0.1.0
     def ==(other)
       self.class == other.class &&
-         self.id == other.id
+        id == other.id
```

Comment:
@jodosha so following should be `true`?
```ruby
Thing.new(id: 12, name: 'Another thing') == Thing.new(id:12, name: 'A thing')
```

I think the problem lies within the multiple meanings of `equal`

* Same state
* Same entity
* And the combination

From a user point of view i'd expect the above code sample toe be `false`